### PR TITLE
Finish PlayerTemplate and PlayerList, complete the definition of Player and implement some Player functions

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/game/common/rts/playertemplate.h [hunk 1]
```diff
@@ -114,6 +139,7 @@ class PlayerTemplateStore : public SubsystemInterface
     PlayerTemplate *Find_Player_Template(NameKeyType key);
     PlayerTemplate *Get_Nth_Player_Template(int index);
     void Get_All_Side_Strings(std::list<Utf8String> *list);
+    int Get_Player_Template_Count() { return m_playerTemplates.size(); }
```
Comment (xezon) [inline src/game/common/rts/playertemplate.h]: Can be const.

File: src/game/common/rts/playertemplate.h [hunk 2]
```diff
@@ -49,10 +49,35 @@ class PlayerTemplate
     Image *Get_Side_Icon_Image();
     Image *Get_General_Image();
     Image *Get_Enabled_Image();
+    Utf8String Get_Starting_Unit(int unit) const;
+
     Utf8String Get_Name() { return g_theNameKeyGenerator->Key_To_Name(m_nameKey); }
     Utf8String Get_Side_Name() { return m_side; }
-    bool Check_Name_Key(NameKeyType key) { return key == m_nameKey; }
+    NameKeyType Get_Name_Key() { return m_nameKey; }
```
Comment (xezon) [inline src/game/common/rts/playertemplate.h]: Can be const.

File: src/game/common/rts/player.cpp [hunk 3]
```diff
@@ -145,28 +209,28 @@ Team *Player::Get_Default_Team()
...
 {
 #ifdef GAME_DLL
-    Call_Method<void, Player, bool>(PICK_ADDRESS(0x00456820, 0x0086048D), this, b);
+    Call_Method<void, Player, bool>(PICK_ADDRESS(0x00456820, 0x0086048D), this, change);
 #endif
 }
 
 // zh: 0x00450AA0, wb: 0x0085A70A
-bool Player::Remove_Team_Relationship(const Team *team)
+bool Player::Remove_Team_Relationship(const Team *that)
```
Comment (xezon) [inline src/game/common/rts/player.cpp:220]: `team` is more descriptive than `that`.

File: src/game/common/rts/player.cpp [hunk 4]
```diff
@@ -315,3 +382,47 @@ void Player::Iterate_Objects(void (*func)(Object *, void *), void *data) const
...
+    Team *team = g_theTeamFactory->Find_Team(str);
+    captainslog_dbgassert(team != nullptr, "no team");
+
+    if (team != nullptr) {
+        m_defaultTeam = team;
+        team->Set_Active();
+    }
+}
+
+void Player::Set_Player_Relationship(const Player *that, Relationship r)
```
Comment (xezon) [inline src/game/common/rts/player.cpp:416]: Can call it `player` instead of `that`.

`that` has no meaning.

File: src/game/common/rts/player.cpp [hunk 5]
```diff
@@ -13,24 +13,88 @@
...
+
+    for (auto i = m_playerTeamPrototypes.begin(); i != m_playerTeamPrototypes.end(); i++) {
+        (*i)->Set_Owning_Player(nullptr);
+    }
+
+    m_playerTeamPrototypes.clear();
+    m_teamRelations->Delete_Instance();
+    m_playerRelations->Delete_Instance();
+
+    for (int i = 0; i < 10; i++) {
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: ARRAY_SIZE(m_squads)

File: src/game/common/rts/player.cpp [hunk 6]
```diff
@@ -13,24 +13,88 @@
...
+{
+    m_playerRelations = new PlayerRelationMap();
+    m_teamRelations = new TeamRelationMap();
+    m_activeBattlePlans[0] = 0;
+    m_activeBattlePlans[1] = 0;
+    m_activeBattlePlans[2] = 0;
+    m_side = nullptr;
+    m_baseSide = nullptr;
+
+    for (int i = 0; i < 10; i++) {
```
Comment (xezon) [inline src/game/common/rts/player.cpp]: ARRAY_SIZE(m_squads)

File: src/game/common/rts/player.cpp [hunk 7]
```diff
@@ -13,24 +13,88 @@
...
+    m_canBuildUnits(true),
+    m_canBuildBase(true),
+    m_bountyCostToBuild(0.0f),
+    m_playerColor(0),
+    m_aiSquad(nullptr),
+    m_rankLevel(0),
+    m_sciencePurchasePoints(0),
+    m_currentSkillPoints(0),
+    m_lastAttackedByFrame(0),
+    m_unitsShouldHunt(0)
```
Comment (xezon) [inline src/game/common/rts/player.cpp:48]: Are the rest of inits covered in `Init()` function? Does not look like much here.

File: src/game/logic/ai/squad.h [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+public:
+    virtual ~Squad() override;
+    virtual void CRC_Snapshot(Xfer *xfer) override;
+    virtual void Xfer_Snapshot(Xfer *xfer) override;
+    virtual void Load_Post_Process() override;
+
+private:
+    std::vector<ObjectID> m_objectIDs;
+    std::vector<Object *> m_objectsCached;
+};
```
Comment (xezon) [inline src/game/logic/ai/squad.h]: eol

File: src/game/common/rts/playertemplate.h (hunk 2 above)
Comment (jonwil) [reply src/game/common/rts/playertemplate.h]: fixed.

File: src/game/common/rts/playertemplate.h (hunk 1 above)
Comment (jonwil) [reply src/game/common/rts/playertemplate.h]: fixed.

File: src/game/common/rts/player.cpp (hunk 7 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp:48]: Yes.

File: src/game/logic/ai/squad.h (hunk 8 above)
Comment (jonwil) [reply src/game/logic/ai/squad.h]: fixed.

File: src/game/common/rts/player.cpp (hunk 5 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

File: src/game/common/rts/player.cpp (hunk 6 above)
Comment (jonwil) [reply src/game/common/rts/player.cpp]: fixed.

**Comment by xezon** [comment]
There is something wrong with teams. Team mates do not show properly. Hidden in fog.

## This commit

shot_20230223_234547_3

## Previous commit

shot_20230223_235616_1

**Review by xezon — COMMENTED** [review]
see above

**Comment by jonwil** [comment]
Issue with not being able to see allies is fixed.

**Review by xezon — COMMENTED** [review]
Hit crash on Defcon 6 (6) map unload in Skirmish with 2 Med AI teammates and 3 Med AI enemies.

Match was played for a couple minutes to build bases. SAVE GAME functionality was attempted to be used twice.

```
 	kernel32.dll!77c5ee9b()	Unknown
 	kernel32.dll![Frames below may be incorrect and/or missing, no symbols loaded for kernel32.dll]	Unknown
 	game.dat!004a5c29()	Unknown
 	game.dat!004a2f01()	Unknown
 	game.dat!004f88cd()	Unknown
 	game.dat!004f8bfe()	Unknown
 	game.dat!004f891d()	Unknown
 	game.dat!005fa49f()	Unknown
 	game.dat!005ec919()	Unknown
 	game.dat!00614e85()	Unknown
 	game.dat!004a66d9()	Unknown
 	game.dat!004a6d93()	Unknown
>	thyme.dll!GameEngine::Execute() Line 556	C++
 	thyme.dll!Game_Main(int argc, char * * argv) Line 41	C++
 	thyme.dll!main(int argc, char * * argv) Line 703	C++
 	thyme.dll!Main_Func(HINSTANCE__ * hInstance, HINSTANCE__ * hPrevInstance, char * lpCmdLine, int nCmdShow) Line 728	C++
 	game.dat!008e08ac()	Unknown
 	kernel32.dll!77c600f9()	Unknown
 	ntdll.dll!77da7bbe()	Unknown
 	ntdll.dll!77da7b8e()	Unknown
 	game.dat!006e0069()	Unknown
 	game.dat!006e0069()	Unknown
 	game.dat!006e0069()	Unknown
 	game.dat!006e0069()	Unknown
 	game.dat!006e0069()	Unknown
 	game.dat!00650052()	Unknown
 	game.dat!00650052()	Unknown
 	game.dat!006e0069()	Unknown
 	game.dat!006e0069()	Unknown
 	game.dat!006e0069()	Unknown
```

`Exception thrown at 0x77C5EE9B (kernel32.dll) in game.dat: 0xC0000005: Access violation writing location 0x000207E7.`

```
EAX = FFFFFFFF EBX = 00000000 ECX = 000207E7 EDX = 03C677FC ESI = 048280C4 EDI = 77C5EE90 EIP = 7A82BE8F ESP = 0019FA3C EBP = 0019FA50 EFL = 00000000 

0x7A9E5008 = 00A2B684 
```

**Review by xezon — APPROVED** [review]
Tested Skirmish and Network match with and against AI. Tested USA Campaign Mission 1. Tested mass faction unit test map with big battles. Tested Shell Map. Tested in RelWithDebInfo. No unusual new issues with this change.

Looks good.

**Pull request merged**